inference-engine/src/mkldnn_plugin/nodes/ctc_loss.cpp

```diff
@@ -72,257 +74,203 @@ class CTCLossImpl : public ExtLayerBase {
             outputs[0]->getTensorDesc().getBlockingDesc().getOffsetPadding();
 
         const auto& logitsShape = inputs[0]->getTensorDesc().getDims();
-        const auto batchNum = logitsShape[0];
-        const auto maxTime = logitsShape[1];
-        const auto classesNum = logitsShape[2];
+        const size_t batchNum = logitsShape[0];
+        const size_t maxTime = logitsShape[1];
+        const size_t classesNum = logitsShape[2];
 
         int blankIndex = classesNum - 1;
         if (inputs.size() > 4) {
             blankIndex = inputs[4]->cbuffer().as<const int*>()[0];
         }
 
-        std::vector<int> targetD(maxTime);
+        std::vector<int> decodedTargetLenB(batchNum, 0);
+        std::vector<std::vector<int>> targetDB(batchNum);
+        std::vector<std::vector<std::vector<float>>> logProbabilitiesB(batchNum);
+        size_t workAmount2 = 0lu;
+
+        auto threadBody_1 = [&](const int ithr, const int nthr) {
+            size_t start(0lu), end(0lu);
+            splitter(batchNum, nthr, ithr, start, end);
+            if (start >= end)
+                return;
+
+            for (size_t b = start; b < end; b++) {
+                if (logitsLength[b] < 0 || labelsLength[b] < 0 || logitsLength[b] > maxTime || labelsLength[b] > logitsLength[b]) {
+                    std::string errorMsg = _logPrefix + ". Logit length cannot be greater than max sequence length. "
+                        + "Label length cannot be greater than a logit length"
+                        + " and both cannot be negative.\nMaxSeqLen: "
+                        + std::to_string(maxTime) + "; Logit len: " + std::to_string(logitsLength[b])
+                        + "; Label len: " + std::to_string(labelsLength[b]);
+                    errorMsg.copy(resp->msg, sizeof(resp->msg) - 1);
```

Comment:
Is not best solution to modify same variables form threads. Data race can happen.

Please move all checks and buffer preparation out of parallel section.

Reply:
fixed